internal/timeline/timeline.go

```diff
@@ -0,0 +1,195 @@
+package timeline
+
+import (
+	"fmt"
+	"log"
+	"sync"
+)
+
+const EPSILON TIME = 0.0001
+const DEFAULT_TRANSITION TIME = 0.5
+
+var debugTimeline bool = false
+var idSeed int64
+var idSeedMutex sync.Mutex
+
+type ON_STEP func(*TIME) ITimeline
+type ON_ACTIVE func(bool)
+
+type ITimeline interface {
+	GetTimeline() *Timeline
+	Step(time *TIME) ITimeline
+	SetActive(bool)
+}
+
+type Timeline struct {
+	ID           int64
+	active       bool
+	offset       TIME
+	speed        float64
+	fadeIn       *Interval
+	fadeOut      *Interval
+	freezingTime TIME
+	next         ITimeline
+	group        *TimelineGroup
+	onStep       ON_STEP
+	onActive     ON_ACTIVE
+}
+
+func (t *Timeline) Init() *Timeline {
+	idSeedMutex.Lock()
+	defer idSeedMutex.Unlock()
+	t.ID = idSeed
+	idSeed++
+	if debugTimeline {
+		log.Printf("Timeline %X", t.ID)
+	}
+	t.speed = 1.0
+	return t
+}
+
+func (t *Timeline) GetTimeline() *Timeline {
+	return t
+}
+
+func (t *Timeline) SetActive(on bool) {
+	if t.active == on {
+		return
+	}
+	if debugTimeline {
+		log.Println("SetActive", on, t.ID)
+	}
+	t.active = on
+	if t.onActive != nil {
+		t.onActive(on)
+	}
+}
+
+func (t *Timeline) Step(time *TIME) ITimeline {
+	if *time < 0.0 {
+		*time = 0.0
+	}
+
+	var running ITimeline = t
+	var LOOP_LIMIT int = 10000
+	for running != nil && *time > EPSILON && LOOP_LIMIT > 0 {
+		LOOP_LIMIT--
+		var realStep, scaledTime, oldScaledTime TIME
+		var oldRunning ITimeline
+		r := running.GetTimeline()
+
+		// step until time <= 0 || offset <= 0
+		var min TIME = *time // minimum step in consideration of time, offset, fadeout.end, freezingTime
+		if r.offset > EPSILON {
+			if min > r.offset {
+				min = r.offset
+			}
+			*time -= min
+			r.offset -= min
+			continue
+		}
+
+		r.SetActive(true)
+
+		if r.fadeOut != nil {
+			end := r.fadeOut.End()
+			if end < 0.0 {
+				end = 0.0
+			}
+			if min > end {
+				min = end
+			}
+		}
+
+		// step until time <= 0 || fadeOut.end <= 0 || freezingTime <= 0
+		if r.freezingTime > EPSILON {
+			if min > r.freezingTime {
+				min = r.freezingTime
+			}
+			*time -= min
+			r.freezingTime -= min
+			if r.fadeIn != nil {
+				r.fadeIn.Step(min)
+			}
+			if r.fadeOut != nil {
+				r.fadeOut.Step(min)
+			}
+
+			goto CHECK_FADE_OUT
```

Comment:
Please do not use goto.

Reply:
In most cases, the use of goto should be avoided. The exception is when it is the only clear and concise way to handle control flow. By ensuring that all cases ultimately check for a fade-out condition, the control flow remains clear and understandable.